core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -11702,21 +11708,36 @@ protected boolean isUserIdGeneratedByUserStore(String userName, Map<String, Stri
      */
     public String getUserIDFromUserName(String userName) throws UserStoreException {
 
+        String userID;
         UserStore userStore = getUserStore(userName);
         if (userStore.isRecurssive()) {
             return ((AbstractUserStoreManager) userStore.getUserStoreManager())
                     .getUserIDFromUserName(userStore.getDomainFreeName());
         }
 
         if (isUniqueUserIdEnabledInUserStore(userStore)) {
-            return doGetUserIDFromUserNameWithID(userStore.getDomainFreeName());
+            userID = AbstractUserStoreManagerCache.getInstance().getValueFromCache(userName,
+                    RESOLVE_USER_ID_FROM_USER_NAME_CACHE_NAME);
+            if (StringUtils.isEmpty(userID)){
+                return doGetUserIDFromUserNameWithID(userStore.getDomainFreeName());
+            } else {
+                AbstractUserStoreManagerCache.getInstance().addToCache(userName, userID,
```

Comment:
why this is in else statement? We should add to cache just after getting from DB.

Reply:
fixed